runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/DistributionData.java

```diff
@@ -17,48 +17,283 @@
  */
 package org.apache.beam.runners.core.metrics;
 
-import com.google.auto.value.AutoValue;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
 import java.io.Serializable;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Function;
 import org.apache.beam.sdk.metrics.DistributionResult;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.annotations.VisibleForTesting;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.ImmutableMap;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.ImmutableSet;
+import org.apache.datasketches.memory.Memory;
+import org.apache.datasketches.quantiles.DoublesSketch;
+import org.apache.datasketches.quantiles.DoublesSketchBuilder;
+import org.apache.datasketches.quantiles.DoublesUnion;
+import org.apache.datasketches.quantiles.DoublesUnionBuilder;
+import org.apache.datasketches.quantiles.UpdateDoublesSketch;
 
 /**
  * Data describing the the distribution. This should retain enough detail that it can be combined
  * with other {@link DistributionData}.
  *
+ * <p>Datasketch library is used to compute percentiles. See {@linktourl
+ * https://datasketches.apache.org/}.
+ *
  * <p>This is kept distinct from {@link DistributionResult} since this may be extended to include
  * data necessary to approximate quantiles, etc. while {@link DistributionResult} would just include
  * the approximate value of those quantiles.
  */
-@AutoValue
-public abstract class DistributionData implements Serializable {
+public class DistributionData implements Serializable {
+  // k = 256 should yield an approximate error ε of less than 1%
+  private static final int SKETCH_SUMMARY_SIZE = 256;
 
-  public abstract long sum();
+  private final Set<Double> percentiles;
+  private long sum;
+  private long count;
+  private long min;
+  private long max;
+  private Map<Double, Double> percentileResults;
+  private transient Optional<UpdateDoublesSketch> sketch;
+  private transient Map<Double, Function<DoublesSketch, Double>> percentileExtractors;
```

Comment:
Having this map seems quite complicated. Can we use the same set of percentiles instead, and then use the sketch directly to calculate the percentile value to generate the result map?

Reply:
Simplified this by doing a simple conversion.